pageserver/src/tenant/storage_layer.rs

```diff
@@ -137,17 +140,34 @@ pub trait Layer: Send + Sync {
     fn is_in_memory(&self) -> bool;
 
     /// Iterate through all keys and values stored in the layer
-    fn iter(&self) -> Box<dyn Iterator<Item = Result<(Key, Lsn, Value)>> + '_>;
+    #[allow(clippy::type_complexity)]
+    fn iter(&self) -> Result<Box<dyn Iterator<Item = Result<(Key, Lsn, Value)>> + '_>>;
 
     /// Iterate through all keys stored in the layer. Returns key, lsn and value size
     /// It is used only for compaction and so is currently implemented only for DeltaLayer
-    fn key_iter(&self) -> Box<dyn Iterator<Item = (Key, Lsn, u64)> + '_> {
-        panic!("Not implemented")
+    fn key_iter(&self) -> Result<Box<dyn Iterator<Item = (Key, Lsn, u64)> + '_>> {
+        anyhow::bail!("Not implemented")
     }
 
     /// Permanently remove this layer from disk.
     fn delete(&self) -> Result<()>;
 
     /// Dump summary of the contents of the layer to stdout
     fn dump(&self, verbose: bool) -> Result<()>;
+
+    fn downcast_remote_layer(self: Arc<Self>) -> Option<std::sync::Arc<RemoteLayer>> {
```

Comment:
NIT (not relevant to this PR a lot): All those `bail!` and now downcasts look like it's wort to revive the initiative to split layers into better typed enum variants.